File: web/frontend/src/utils/permissions.ts
/**
 * Permission hierarchy utilities for frontend role-based access control.
 * 
 * Matches backend permission levels from core/dependencies.py
 */

export type RoleLevel = 'bot_owner' | 'bot_admin' | 'discord_manager' | 'moderator' | 'staff' | 'user';

/**
 * Role hierarchy values (higher = more privilege).
 * Must match backend ROLE_HIERARCHY in core/dependencies.py.
 */
export const ROLE_HIERARCHY: Record<RoleLevel, number> = {
  bot_owner: 6,
  bot_admin: 5,
  discord_manager: 4,
  moderator: 3,
  staff: 2,
  user: 1,
};

/**
 * Check if a user's role meets the minimum required level.
 * 
 * @param userRole - The user's current role level
 * @param requiredRole - The minimum required role level
 * @returns true if user has sufficient permissions
 * 
 * @example
 * hasPermission('bot_admin', 'moderator') // true (bot_admin > moderator)
 * hasPermission('staff', 'moderator') // false (staff < moderator)
 */
export function hasPermission(userRole: RoleLevel, requiredRole: RoleLevel): boolean {
  const userLevel = ROLE_HIERARCHY[userRole] || 0;
  const requiredLevel = ROLE_HIERARCHY[requiredRole] || 0;
  return userLevel >= requiredLevel;
}

/**
 * Get a user-friendly display name for a role level.
 */
export function getRoleDisplayName(role: RoleLevel): string {
  const displayNames: Record<RoleLevel, string> = {
    bot_owner: 'Bot Owner',
    bot_admin: 'Bot Admin',
    discord_manager: 'Discord Manager',
    moderator: 'Moderator',
    staff: 'Staff',
    user: 'User',
  };
  return displayNames[role] || 'Unknown';
}

/**
 * Get badge color classes for a role level.
 */
export function getRoleBadgeColor(role: RoleLevel): string {
  const colors: Record<RoleLevel, string> = {
    bot_owner: 'bg-purple-900 text-purple-200',
    bot_admin: 'bg-red-900 text-red-200',
    discord_manager: 'bg-orange-900 text-orange-200',
    moderator: 'bg-blue-900 text-blue-200',
    staff: 'bg-green-900 text-green-200',
    user: 'bg-gray-700 text-gray-300',
  };
  return colors[role] || 'bg-gray-700 text-gray-300';
}

/**
 * Get all roles at or above a certain level (for UI filtering).
 */
export function getRolesAtOrAbove(minRole: RoleLevel): RoleLevel[] {
  const minLevel = ROLE_HIERARCHY[minRole];
  return (Object.keys(ROLE_HIERARCHY) as RoleLevel[])
    .filter(role => ROLE_HIERARCHY[role] >= minLevel)
    .sort((a, b) => ROLE_HIERARCHY[b] - ROLE_HIERARCHY[a]); // Descending order
}

File: web/backend/core/permissions.py
from __future__ import annotations

"""
Permission and authorization dependencies for FastAPI routes.

Provides role-hierarchy checking and ``require_*`` dependency factories
consumed by route handlers to enforce access control.
"""

import logging
from typing import TYPE_CHECKING

from fastapi import Cookie, Depends, HTTPException, Request, Response

from .auth import _ensure_fresh_guild_access, get_current_user
from .internal_api_client import InternalAPIClient, get_internal_api_client
from .security import SESSION_COOKIE_NAME

if TYPE_CHECKING:
    from .schemas import UserProfile

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Role hierarchy
# ---------------------------------------------------------------------------

# Higher number = higher privilege.
ROLE_HIERARCHY: dict[str, int] = {
    "bot_owner": 7,
    "bot_admin": 6,
    "discord_manager": 5,
    "moderator": 4,
    "event_coordinator": 3,
    "staff": 2,
    "user": 1,
}


def _has_minimum_role(user_role: str, required_role: str) -> bool:
    """Check if user's role meets minimum requirement.

    Args:
        user_role: User's role level
        required_role: Minimum required role level

    Returns:
        True if user_role >= required_role in hierarchy
    """
    user_level = ROLE_HIERARCHY.get(user_role, 0)
    required_level = ROLE_HIERARCHY.get(required_role, 0)
    return user_level >= required_level


# ---------------------------------------------------------------------------
# Permission dependency factory
# ---------------------------------------------------------------------------


def require_guild_permission(min_role: str):
    """Factory function to create permission dependency for specific role level.

    Checks user's permission level in their active guild against minimum
    requirement. Bot owners always pass regardless of guild membership.

    Args:
        min_role: Minimum required role level (bot_owner, bot_admin,
            discord_manager, moderator, event_coordinator, staff)

    Returns:
        Dependency function that validates permissions

    Example:
        @router.get("/admin-only", dependencies=[Depends(require_guild_permission("bot_admin"))])
    """

    async def check_permission(
        request: Request,
        response: Response,
        raw_session: str | None = Cookie(None, alias=SESSION_COOKIE_NAME),
        current_user: UserProfile = Depends(get_current_user),
        internal_api: InternalAPIClient = Depends(get_internal_api_client),
    ) -> UserProfile:
        """Check if user has required permission level in active guild."""
        from .pagination import is_all_guilds_mode

        await _ensure_fresh_guild_access(
            request,
            response,
            raw_session,
            current_user,
            internal_api,
        )

        if is_all_guilds_mode(current_user.active_guild_id):
            if not current_user.is_bot_owner:
                raise HTTPException(
                    status_code=403,
                    detail="All Guilds mode is only available to bot owners",
                )
            return current_user

        if not current_user.active_guild_id:
            raise HTTPException(status_code=400, detail="No active guild selected")

        guild_perm = current_user.authorized_guilds.get(current_user.active_guild_id)

        if not guild_perm:
            raise HTTPException(
                status_code=403,
                detail=f"Not authorized for guild {current_user.active_guild_id}",
            )

        if not _has_minimum_role(guild_perm.role_level, min_role):
            raise HTTPException(
                status_code=403,
                detail=f"Requires {min_role} role (you have: {guild_perm.role_level})",
            )

        return current_user

    return check_permission


# ---------------------------------------------------------------------------
# Convenience dependency functions for common permission levels
# ---------------------------------------------------------------------------


def require_bot_admin():
    """Require bot_admin level or higher (bot_owner, bot_admin)."""
    return require_guild_permission("bot_admin")


def require_discord_manager():
    """Require discord_manager level or higher."""
    return require_guild_permission("discord_manager")


def require_moderator():
    """Require moderator level or higher."""
    return require_guild_permission("moderator")


def require_event_coordinator():
    """Require event_coordinator level or higher."""
    return require_guild_permission("event_coordinator")


def require_staff():
    """Require staff level or higher."""
    return require_guild_permission("staff")


def require_bot_owner():
    """Require bot_owner level only. Used for privileged operations like leaving guilds."""
    return require_guild_permission("bot_owner")


async def require_is_bot_owner(
    current_user: UserProfile = Depends(get_current_user),
) -> UserProfile:
    """
    Require user to be a bot owner (checked via session is_bot_owner flag).

    Does NOT require an active guild - useful for global operations like bot invite.
    """
    if not current_user.is_bot_owner:
        raise HTTPException(
            status_code=403,
            detail="This action requires bot owner permissions",
        )
    return current_user
